[AIRFLOW-3737] Kubernetes executor cannot handle long dag/task names

PaulW opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Airflow Jira] https://issues.apache.org/jira/browse/AIRFLOW-3737

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:
Kubernetes has a 63char limit on label values, so when running either a
subdag or a dag with a name with a length longer than 63 chars, the pod
creation process will fail with the following error:

  [2019-01-21 08:07:05,337] {rest.py:219} DEBUG - response body:
  {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"unable
  to parse requirement: invalid label value:
  \"very_long_dag_name.very_long_task_name\": must be no more than 63
  characters","reason":"BadRequest","code":400}

Annotations however, allow for a relatively unrestricted length when
storing them against pods, and as such would be better suited in this
situation.

### Tests

- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
This pull request fixes a bug in `KubernetesJobWatcher`, which doesn't yet have any tests
(it is patched out in tests of the `KubernetesExecutor`).  It isn't straightforward to add meaningful
tests to the `KubernetesJobWatcher` due to its integration with `kube_client`. We could pass in
a mock, but we don't think this would actually achieve anything and could actually create difficulties
in the future as we would be making assumptions on how to code works, instead on what the code
does. This is really crying out for an integration test, but we believe that is outside the scope of this
pull request. We're happy to take advice from the community on this 

### Commits

- [X] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [X] Passes `flake8`

PaulW committed:
[AIRFLOW-3737] Kubernetes executor cannot handle long dag/task names

Kubernetes has a 63char limit on label values, so when running either a
subdag or a dag with a name with a length longer than 63 chars, the pod
creation process will fail with the following error:

  [2019-01-21 08:07:05,337] {rest.py:219} DEBUG - response body:
  {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"unable
  to parse requirement: invalid label value:
  \"very_long_dag_name.very_long_task_name\": must be no more than 63
  characters","reason":"BadRequest","code":400}

Annotations however, allow for a relatively unrestricted length when
storing them against pods, and as such would be better suited in this
situation.

ashb commented on airflow/contrib/kubernetes/worker_configuration.py:
```diff
@@ -226,6 +227,9 @@ def make_pod(self, namespace, worker_uuid, pod_id, dag_id, task_id, execution_da
         if gcp_sa_key:
             annotations['iam.cloud.google.com/service-account'] = gcp_sa_key
 
+        annotations['dag_id'] = dag_id
```
Is it worth "namespacing" these annotations to something like `org.apache.airflow.dag_id`?

(I don't know if that's recommended or needed in kube-land)

PaulW replied:
Yes you're likely right.  From the docs, it does mention that automated processes must specify a prefix, so with that I think `org.apache.airflow/dag_id` and `org.apache.airflow/task_id` would probably be preferential.

PaulW committed:
Use a prefix when setting annotations

ashb reviewed (approved):
Changes look sensible to be, but I'm not yet familair with the kube executor.

Could someone else who uses kube in production chime in on this?

pgagnon reviewed (commented):
I feel that this is a bit overkill, in the sense that this will also affect dag and task names with a length < 63 characters. This brings downsides since it will be much harder to select dags by label after this change.

What I would suggest instead of implementing `{dag, task}_hash` this way could be to add a `{dag, task}_slug` instead, which would be an url-safe representation of the `{dag, task}` name. If the slug would end up being > 63 characters, truncate it to, say, 56 characters and add a checksum of the dag name at the end.

example: `<56_character_string>-a2da95ad`

PaulW committed:
Rework patch to no longer require the use of annotations
We now create a safe label value in accordance to the constraints
defined here:
https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/
We then append the label value with a hash if either the length of the
label value is greater than 63chars, or if we have performed any modification
to the label value.

PaulW committed:
Flake8 fix

dimberman commented on airflow/contrib/executors/kubernetes_executor.py:
```diff
@@ -476,6 +480,27 @@ def _make_safe_pod_id(safe_dag_id, safe_task_id, safe_uuid):
+        """
+        MAX_LABEL_LEN = 63
+
+        safe_label = re.sub(r'^[^a-z0-9A-Z]*|[^a-zA-Z0-9_\-\.]|[^a-z0-9A-Z]*$', '', string)
```
so if I'm understanding this correctly this will prevent any collisions. So if I have "my*favorite*dag" and "my^favorite^dag", then one will be "myfavoritdag<hash1>" and the second will be  "myfavoritedag<hash2>". Is that correct?

PaulW replied:
Yes that is correct.  As the regex and formatting syntax on the kubernetes docs is quite precise, if we were to simply hash only if the modified labels were above 63 chars then there is a strong possibility that it could lead to collisions if the label length is less than 63 chars, and the unsupported chars that were removed being the only difference between the dags (as you've mentioned above).  This also covers accentuated characters if they are used in dag names too.

Therefor, in your example, for your first dag "_myfavoritedag_" this would have the label value set as "_myfavoritedag_", whereas the second dag "_my^favorite^dag_" would be set as "_myfavouritedag-99a5a18ff_" due to the fact that it is modified compared to the original value.

ashb reviewed (changes requested):
I'd like to see tests covering _make_safe_label_value please.

XD-DENG commented on airflow/contrib/executors/kubernetes_executor.py:
```diff
@@ -477,6 +481,27 @@ def _make_safe_pod_id(safe_dag_id, safe_task_id, safe_uuid):
+
+        If the label value is then greater than 63 chars once made safe, or differs in any
+        way from the original value sent to this function, then we need to truncate to
+        53chars, and append it with a unique hash.
```
`53chars` is a typo?

PaulW replied:
No that's correct.  If we are hashing the string after it's been modified in any way, we need to truncate the result from `AirflowKubernetesScheduler._make_safe_label_value()` to a maximum of 53chars, as we are adding a 9 char has to the end.  If we were to truncate to 63chars, when the has is added, it would then make the string 73chars max length, which would be too long for the label value length limit.

XD-DENG reviewed (dismissed):
Potential typo.

PaulW committed:
Squashed commit of the following:

commit 8a3cfd9025fd9c504b4a25fb1d8a6ef3cc2f8a3f
Author: Paul.Bramhall <paul.bramhall@autotrader.co.uk>
Date:   Fri Mar 22 10:16:28 2019 +0000

    flake8 fix

commit 0e22fb46e83c0cc63122b32adef5d9f0a7b5e3c3
Author: Paul.Bramhall <paul.bramhall@autotrader.co.uk>
Date:   Fri Mar 22 09:59:19 2019 +0000

    Test an explicit non-randomised test as requested by ashb

commit c124589864405b18d8b863e93c6e23ac601af820
Author: Paul.Bramhall <paul.bramhall@autotrader.co.uk>
Date:   Tue Mar 19 15:30:47 2019 +0000

    Generate deterministic random data for dag_id and task_id
    To ensure testing consistency, when calling the '_gen_random_string()'
    function, we want to set the seed dependant on the string length and
    iteration within the `cases.extend` loop.
    This allows random values to be created (as required) but ensuring
    consistency of these values during each and every time we run tests
    against the code.

commit 0daa0ec301512a9822935c24a03a6a026c826b77
Author: Paul.Bramhall <paul.bramhall@autotrader.co.uk>
Date:   Tue Mar 19 11:27:29 2019 +0000

    Add tests around 'make_safe_label_value'
